src/meltano/cli/compile.py

```diff
@@ -0,0 +1,90 @@
+"""CLI command `meltano compile`."""
+
+from __future__ import annotations
+
+import json
+from pathlib import Path
+
+import click
+
+from meltano.cli import cli
+from meltano.cli.params import pass_project
+from meltano.cli.utils import CliError, InstrumentedCmd
+from meltano.core.environment import Environment
+from meltano.core.environment_service import EnvironmentService
+from meltano.core.manifest import Manifest
+from meltano.core.project import Project
+from meltano.core.tracking import Tracker
+from meltano.core.tracking.contexts import CliEvent
+
+
```

Comment:
Per the engineering all-hands discussion on 2023-01-17: to allow for a transition from old non-compliant `meltano.yml` files to compliant ones, we should make validation failures into a warning rather than an error until Meltano 3.0. These warnings should state that the checked file is invalid, and will not work in Meltano 3.0. Bonus points if a user-friendly error message (e.g. "the key "arbitrary-key" on line 123 is not permitted") from the validation error can be extracted an displayed within the warning message.

Reply:
If we try to perform this validation using the `jsonschema` Python package then we need to know which validator class to use ahead of time (even though our choice of validator should be based on the value of the `$schema` field within the schema file), and we need to extract a user-friendly error message (which seems difficult).

Instead, I think we should add the CLI tool `check-jsonschema` as a Meltano dependency, and call it in a subprocess. This tool comes recommended by the maintainers of `jsonschema` as a replacement for the deprecated `jsonschema` CLI, and it provides clear error messages when a validation fails.

Depending on a CLI tool like this is not the most elegant solution, particularly since it uses the `jsonschema` library internally, but this is probably the best path forwards in terms of value ÷ effort, and I don't think this will harm maintainability.

Reply:
I've implemented this by calling `check-jsonschema` in a subprocess. Because the merged project files only exist in-memory, I wrote them to a temporary file. This leads to error messages from `check-jsonschema` like:

```
Schema validation errors were encountered.
  /tmp/tmpytbqo3c1::$: Additional properties are not allowed ('bad_key' was unexpected)
```

Where `/tmp/tmpytbqo3c1` no longer exists once the message is printed to console. We can improve the debugging experience here by storing the merged project files within something like `.meltano/project.json` (with an interprocess lock to ensure only one process is accessing that particular path at a time). We should also include whitespace in the JSON file we generate to make it more readable, and to make the error messages from `check-jsonschema` include line and column numbers.

One other concern is colouring. By default `check-jsonschema` will colour its output, but when run from a subprocess under Meltano it doesn't. We'll want it to provide colourful output unless `NO_COLOR` (or something equivalent) is set.

Reply:
`check-jsonschema` doesn't identify line or column numbers, so maybe all we need on that front is for it to say `meltano.yml` rather than `/tmp/asdfqwerty` (which won't exist if the user checks).

Using `check-jsonschema` as a library to bypass Click's auto-stripping of ANSI color codes has so far not been successful. We need to capture the output, which it sends to stdout, but if we use `contextlib.redirect_stdout` then Click strips the color codes.

I've opened a PR in `check-jsonschema` to have Click respect its internal color setting, but it seems like that color setting is actually entirely unused, and does not get set to `False` when `$NO_COLOR` is set (despite `check-jsonschema` explicitly supporting the `NO_COLOR` env var).

Reply:
I have closed the PR linked above, and superseded it with:
- https://github.com/python-jsonschema/check-jsonschema/pull/227

Using the changes from that PR, we get output like the following:

![image](https://user-images.githubusercontent.com/11428666/213749024-56e46c26-adc1-4354-9afc-83129c8e5777.png)

One warning because the project files violate the schema, and another because the newly compiled manifest file violates the schema.